File: c/pangram/pangram.c
#include "pangram.h"

#include <ctype.h>
#include <string.h>

#define TWENTY_SIX_ONE_BITS 67108863

bool is_pangram(const char *sentence)
{
    if (sentence == NULL) {
        return false;
    }

    /*
     * Mask the bits corresponding to each letter's position in the alphabet.
     */
    int mask = 0;
    for (char c; (c = tolower(*sentence)) != '\0'; sentence++) {
        if (isalpha(c)) {
            mask |= 1 << (c - 'a');
        }
    }

    return mask == TWENTY_SIX_ONE_BITS; // Are all 26 bits masked?
}

File: go/pangram/pangram.go
package pangram

import "unicode"

const twentySixOneBits = 67108863

func IsPangram(input string) bool {
	if len(input) < 26 {
        return false
    }

    mask := 0
    for _, c := range input {
        c = unicode.ToLower(c)
        if unicode.IsLetter(c) {
            mask |= 1 << (c - 'a')
        }
    }

    return mask == twentySixOneBits
}
